deepmarket/PLUTO
# Docker script

Adding the ability for PLUTO to spin up a docker container when a resource is added.  Quick summary of the changes:

1. `dashboard.py` updating response object to use 'data' instead of 'resources' in response to API updates
2. `resource_controller_ui.py` updating table to have column for docker container ID
3. `resources.py` added ability to spin up a container when resource is added, reformatted a bit of the how the machine info is obtained to add a bit of error handling
4. `util.py` added docker functionality

**To test**
after spinning up PLUTO, run through the add resources workflow, then make sure that a container is created.  In the new docker container id column, a resource that's been added should have the short form ID.  To check that there is a running container, from a terminal run `docker ps` and make sure that there is a container with a matching ID running the `samgomena/deepshare_worker` image.  You can check that the RAM / CPU usage that you've entered is correct by running `docker inspect [container id] | grep -i cpu` and looking for the [nano_cpus](https://docs.docker.com/engine/api/v1.40/#operation/ContainerCreate) variable. You can verify the RAM given to the container by doing a similar process: `docker inspect [container id] | grep -i memory` and verifying the "memory" variable.

When removed, the container should stop running, and `docker ps` should not be able to find it. 

**Note**
At this point it is not possible to create multiple containers on the same host as they will both have the same IP address.
